app/scripts/controllers/network/network-controller.js

```diff
@@ -179,62 +197,106 @@ export default class NetworkController extends EventEmitter {
     if (EIPS[1559] !== undefined) {
       return EIPS[1559];
     }
-    const latestBlock = await this._getLatestBlock();
-    const supportsEIP1559 =
-      latestBlock && latestBlock.baseFeePerGas !== undefined;
-    this._setNetworkEIPSupport(1559, supportsEIP1559);
+    const supportsEIP1559 = await this._determineEIP1559Compatibility();
+    this.networkDetails.putState({
+      EIPS: {
+        ...this.networkDetails.getState().EIPS,
+        1559: supportsEIP1559,
+      },
+    });
     return supportsEIP1559;
   }
 
+  /**
+   * Captures information about the currently selected network — namely,
+   * the network ID and whether the network supports EIP-1559 — and then uses
+   * the results of these requests to determine the status of the network.
+   */
   async lookupNetwork() {
-    // Prevent firing when provider is not defined.
+    const { chainId } = this.providerStore.getState();
+    let networkChanged = false;
+    let networkId;
+    let supportsEIP1559;
+    let networkStatus;
+
     if (!this._provider) {
       log.warn(
         'NetworkController - lookupNetwork aborted due to missing provider',
       );
       return;
     }
 
-    const { chainId } = this.providerStore.getState();
     if (!chainId) {
       log.warn(
         'NetworkController - lookupNetwork aborted due to missing chainId',
       );
-      this._setNetworkState('loading');
-      this._clearNetworkDetails();
+      this._resetNetworkId();
+      this._resetNetworkStatus();
+      this._resetNetworkDetails();
       return;
     }
 
-    // Ping the RPC endpoint so we can confirm that it works
-    const initialNetwork = this.networkStore.getState();
-    const { type } = this.providerStore.getState();
-    const isInfura = INFURA_PROVIDER_TYPES.includes(type);
-
-    if (isInfura) {
-      this._checkInfuraAvailability(type);
-    } else {
-      this.emit(NETWORK_EVENTS.INFURA_IS_UNBLOCKED);
-    }
+    this.once(NETWORK_EVENTS.NETWORK_DID_CHANGE, () => {
+      networkChanged = true;
+    });
 
-    let networkVersion;
-    let networkVersionError;
     try {
-      networkVersion = await this._getNetworkId();
+      const results = await Promise.all([
+        this._getNetworkId(),
+        this._determineEIP1559Compatibility(),
+      ]);
+      networkId = results[0];
+      supportsEIP1559 = results[1];
+      networkStatus = NetworkStatus.Available;
     } catch (error) {
-      networkVersionError = error;
+      let data;
+      if (hasProperty(error, 'data')) {
+        data = error.data;
+      } else if (hasProperty(error, 'code')) {
+        try {
+          data = JSON.parse(error.message);
+        } catch {
+          // error.message must not be JSON for some reason
+        }
+      }
+      if (data) {
```

Comment:
Ah; I see that our Infura middleware is responsible for this odd message-JSON situation. If there is an unexpected HTTP error response, the _text_ response body is used as the message.

There is no such case in the fetch middleware we use for custom networks though.

Maybe we should split these two cases, and only attempt parsing the message as JSON for Infura networks? We only want to set this blocked status for Infura networks anyway.

Reply:
But about this condition, putting the Infura blocked case aside, it should be safe to consider a standard non-internal JSON-RPC error as indication that the network is _unavailable_.

The "unknown" case is tricker though because there are categories of error that could indicate either a lack of availability or an unanticipated error in our stack. For example we return "internal" errors and non-RPC errors today in various network error situations, despite the fact that these errors should indicate an error in _our code_.

But I'd be OK with treating these cases as "unknown" for now anyway, and working towards improving error handling in the future to make that distinction more meaningful. With the exception of the Infura blocked error, which is an "internal" error right now but we know to expect it.

We could do that by checking the code of the error. If the code doesn't exist, or if it's `-32603` (for "internal"), then we consider it to be "unknown" (unless we detect that it's a "blocked" response). Otherwise we can consider it to be "unavailable".

Does that sound reasonable? Are there alternatives we could consider? One alternative would be to use "unknown" just for the initial state and consider anything else "unavailable". That would be closer to what we do today, but seems less helpful for debugging situations.

Reply:
Good point. At the end of the day, we are trying to determine whether `countryBlocked` is in the response. I think I was trying to look at both the Infura middleware and fetch middleware to see how they represented errors, but we only really care about the Infura middleware. In that case, [all it does is call `ethErrors.rpc.internal` with the unencoded JSON response body](https://github.com/MetaMask/eth-json-rpc-infura/blob/9b6dc41aaa9226fa4a382896d0aee38ec6268e73/src/create-infura-middleware.ts#L173), which will then [create an error object passing along the body as the `message`](https://github.com/MetaMask/eth-rpc-errors/blob/c3ea9e521aa75c553802080b44ff665e428d00cd/src/errors.ts#L64). So I can clean this up a bit.

Reply:
I didn't see your last two comments until now, but is more of what you had in mind? [3f0f13d](https://github.com/MetaMask/metamask-extension/pull/17556/commits/3f0f13d1b2349f00bd19afcc9963fa4ba1383386)

Reply:
Yes, definitely closer, except I'm not sure we should be using `instanceof` here because we have multiple versions of `eth-rpc-errors` in play. The extension is using a different version than the middleware, so it would always be `false` until they were aligned.

Checking the `code` property should be reliable across versions though.

There is also the "internal error" handling to consider; I was suggesting that we should treat internal errors as "unknown" because they should indicate (as the name implies) errors internal to our codebase rather than network errors. But I don't care strongly either way there because of how we're misusing internal errors today.

Reply:
Okay, it sounds like with my solution, we are always treating an RPC error as unavailable if it's not blocked. I'm also making a check for EthereumRpcError. This may not be necessary, but I'll try adding another check for `code` on top of what I have and see what you think.

Reply:
Following up on this, I pushed an update that uses `error.code` to test for an RPC error in [908c366](https://github.com/MetaMask/metamask-extension/pull/17556/commits/908c36699c58160ff7bdd8bff4d3717f7fd47cc1) and added a check for internal errors in [489ae5c](https://github.com/MetaMask/metamask-extension/pull/17556/commits/489ae5cdf01ade3ebdce458192c692f7cc8ef99f).

Reply:
Thanks, this looks really great now